src/bezier.cpp

```diff
@@ -560,94 +519,100 @@ PointVector Curve::pointsOfIntersection(const Curve& curve, bool stop_at_first,
   return points_of_intersection;
 }
 
-double Curve::projectPoint(const Point& point, double step, double epsilon, std::size_t max_iter) const
+Parameter Curve::projectPoint(const Point& point, double epsilon) const
 {
-  step = std::max(step, 0.01);
-  epsilon = std::max(epsilon, 0.001);
+  /* Chen, X. D., Zhou, Y., Shu, Z., Su, H., & Paul, J. C. (2007, August). Improved Algebraic Algorithm on Point
+   * projection for Bezier curves. In Second International Multi-Symposiums on Computer and Computational Sciences
+   * (IMSCCS 2007) (pp. 158-163). IEEE.
+   */
 
-  double t = 0;
-  double t_dist = (valueAt(t) - point).norm();
-
-  // Coarse search
-  for (double k = step; k < 1 + step; k += step)
+  if (!cached_projection_polynomial_part_)
   {
-    double new_dist = (valueAt(k) - point).norm();
-    if (new_dist < t_dist)
-    {
-      t_dist = new_dist;
-      t = k;
-    }
+    Eigen::MatrixXd curve_polynomial = (bernsteinCoeffs() * control_points_);
+    Eigen::MatrixXd derivate_polynomial = (derivative()->bernsteinCoeffs() * derivative()->control_points_);
+
+    Eigen::VectorXd polynomial_part = Eigen::VectorXd::Zero(curve_polynomial.rows() + derivate_polynomial.rows() - 1);
+    for (uint k = 0; k < curve_polynomial.rows(); k++)
+      for (uint i = 0; i < derivate_polynomial.rows(); i++)
+        polynomial_part(k + i) += curve_polynomial.row(k).dot(derivate_polynomial.row(i));
```

Comment:
Could this be done by multiplying matrices?

Reply:
If one is row-wise (N) and the other is col-wise (N-1), the result would be N x N-1 matrix and a result would then have to be calculated as a sum of all antidiagonals

Reply:
looks like my hunch was right, found a question about it on the forum: https://forum.kde.org/viewtopic.php?f=74&t=97255

Reply:
redone with block operations; line 540 also redone in a similar manner